### PR TITLE
fix(octopus): free session window must be measured from minutes_now

### LINKED ISSUE #4931: Sunday's Power Up events are recognised but not being added to the plan
**Describe the bug**
I have selected my two Octopus happy hour power up events.

They appear in the power up sensor:
image

And in the logfile I can see that Predbat recognises them:
image

However in the plan, Predbat doesn't appear to be have altered the rates and thus is not planning for them:
image

I wonder if this is something to do with the carried forward rates which are all appearing with a '?' suffix, indicating that the rates are not yet present in the Octopus entities and have been assumed from prior day rates.

Today is Friday, the power up events are 2 days away so maybe when the rates appear from Octopus the power up override will then appear, but I would have expected them to be planned for properly and not defaulting to a roll-forward of prior day rates which is what is happening.

**Expected behaviour**
Power up events are planned with a zero import cost.

**Predbat version**
8.54.3

**Environment details**
HAOS 2026.8.3
Givenergy inverters

**Log file**
[predbat.log](https://github.com/user-attachments/files/31848959/predbat.log)

**Predbat debug yaml file**
[1800.txt](https://github.com/user-attachments/files/31849030/1800.txt)

### PR DESCRIPTION
**Pull request by mgazza**
Fixes #4931.

## Root cause

`load_free_slot()` bounded the session against `self.forecast_minutes` on its own:

```python
end_minutes = min(minutes_to_time(end, self.midnight_utc), self.forecast_minutes)

if start_minutes >= 0 and end_minutes != start_minutes and start_minutes < self.forecast_minutes:
```

`self.forecast_minutes` is a **duration** from `minutes_now` (`forecast_hours * 60`, `fetch.py`), while `start_minutes` / `end_minutes` are **absolute** minutes from `midnight_utc`. The two are not comparable. The effective cut-off was 48 hours from *midnight today* rather than 48 hours from *now* — so the accepted window shrank as the day went on.

The two sibling loaders in the same file already get this right:

| loader | bound |
|---|---|
| `load_saving_slot` | `self.forecast_minutes + self.minutes_now` |
| `load_axle_slot` | `self.forecast_minutes + self.minutes_now` |
| **`load_free_slot`** | **`self.forecast_minutes`** |

## Why it presents as "recognised but not planned"

`fetch_octopus_sessions()` logs the event when it fetches it, so the session shows up in the sensor and in the log. The `Setting Octopus free session in range ...` line sits *inside* the failing `if`, so when the slot is dropped there is no second log line and nothing to indicate it was discarded. From the outside the rates just never change — the symptom reported in #4931.

Worked example from that report — Friday 18:00, session Sunday 11:00–12:00:

- `start_minutes` = 3540 (Sunday 11:00 measured from Friday midnight)
- `forecast_minutes` = 2880
- `3540 < 2880` → false → **session discarded**

…even though the plan runs to `minutes_now + forecast_minutes` = 1080 + 2880 = 4320 and covers minute 3540 comfortably. Anyone told about a free hour more than a day or so ahead loses it, and the further into the day they look, the worse it gets.

`rate_replicate()` fills to `forecast_minutes + 48 * 60`, and runs before `load_free_slot()` in `fetch.py`, so the extended range is already populated — no new keys are needed.

## Tests

`test_load_free_slot` previously never set `minutes_now`, so it inherited the wall clock and its boundary cases meant something different depending on when they ran. It now pins `minutes_now` to 18:00 and sizes the rate dicts the way `rate_replicate` does.

- **New Test 6b** — the two-day-ahead session from #4931. Fails without this change: minute 3540 stays at 20.0p.
- **Test 6** — previously used a day-3 session, which is inside the corrected window; moved to a session genuinely past `minutes_now + forecast_minutes`.
- **Test 7** — previously asserted the cap landed on `forecast_minutes`; now asserts it lands on `minutes_now + forecast_minutes`, and that the minute after the window is untouched.

Tests 6 and 7 encoded the old boundary, so they had to move with it — flagging that explicitly rather than burying it.

`unit_test.py -k load_free_slot / octopus_free / saving_session / octopus_slots / savings_stability / fetch_octopus_rates` all pass.

## Noted, not changed

`rate_add_io_slots()` clamps `end_minutes` with a bare `self.forecast_minutes` too. It is left alone here — IO dispatch slots are near-term and changing that behaviour deserves its own change with its own reasoning — but it looks like the same latent mismatch and may be worth a separate look.

## Related

springfall2008/batpred#4930 raises the **manual** rate-override horizon from 48 hours to 7 days. Different code path, same user-facing complaint: a free hour announced several days ahead does not make it into the plan. Regional Power Ups are announced per-customer and are not on the national free-electricity page, so the manual override is the only route for them — which is why both halves matter.

### CODE REVIEW

File: apps/predbat/octopus.py [hunk 1]
```diff
@@ -2853,10 +2853,13 @@ def load_free_slot(self, octopus_free_slots, rate_dict, export=False, rate_repli
...
+                # forecast_minutes is a duration from minutes_now, while start/end are absolute
+                # minutes from midnight_utc - so the window end is minutes_now + forecast_minutes.
+                # load_saving_slot() and load_axle_slot() both bound themselves that way.
                 start_minutes = minutes_to_time(start, self.midnight_utc)
-                end_minutes = min(minutes_to_time(end, self.midnight_utc), self.forecast_minutes)
+                end_minutes = min(minutes_to_time(end, self.midnight_utc), self.forecast_minutes + self.minutes_now)
 
-            if start_minutes >= 0 and end_minutes != start_minutes and start_minutes < self.forecast_minutes:
+            if start_minutes >= 0 and end_minutes != start_minutes and start_minutes < (self.forecast_minutes + self.minutes_now):
                 self.log("Setting Octopus free session in range {} - {} export {} rate {}".format(self.time_abs_str(start_minutes), self.time_abs_str(end_minutes), export, rate))
```
[the remark this answers is not in the document]
Comment (springfall2008) [reply apps/predbat/octopus.py]: Automated reply from the triage bot. Confirmed and fixed in e2f56808: on a decode failure `start_minutes`/`end_minutes` kept the previous slot's values and the unconditional apply block re-applied the new `rate` over the old range (directly observable on the `export=True` path, which writes `rate` outright). The range computation and apply block are now both nested inside the decoded `if start and end:` branch, so a slot that fails to decode is skipped entirely — no stale range, no leftover-`start_minutes = 0` initialisation. Added test 5b to `test_load_free_slot.py` covering good-slot-then-undecodable-slot, which fails against the old code.

File: apps/predbat/tests/test_load_free_slot.py [hunk 2]
```diff
@@ -194,55 +201,81 @@ def test_load_free_slot(my_predbat):
...
+    my_predbat.rate_import = {n: 20.0 for n in range(0, window_end + 24 * 60)}
+    my_predbat.load_scaling_dynamic = {n: 1.0 for n in range(0, window_end + 24 * 60)}
 
-    # Slot ending beyond forecast window
-    partial_slot = [{"start": "2025-01-16T23:00:00+00:00", "end": "2025-01-17T02:00:00+00:00", "rate": 0.0}]  # Near end of day 2  # Extends into day 3 (beyond forecast)
+    # Slot straddling the end of the plan window (minute 4320 = 00:00 on day 4)
+    partial_slot = [{"start": "2025-01-17T23:00:00+00:00", "end": "2025-01-18T02:00:00+00:00", "rate": 0.0}]
 
     rate_replicate = {}
     my_predbat.load_free_slot(partial_slot, my_predbat.rate_import, export=False, rate_replicate=rate_replicate)
```
[the remark this answers is not in the document]
Comment (springfall2008) [reply apps/predbat/tests/test_load_free_slot.py]: Automated reply from the triage bot. Confirmed — `window_end` is `2880 + 1080 = 3960` (18:00 on day 3), so the old slot started at minute 4260, after the window: the cap loop was empty and the `rate_import[window_end]` check passed trivially. Replaced it in e2f56808 with a slot that genuinely straddles the window (17:00–19:00 on day 3), asserting minutes 3900–3960 are set to 0.0 and minute 3960 is untouched. Also corrected the neighbouring comments that cited the phantom 4320 boundary, and added test 5b for the stale-range bug the other thread flagged.
